lib/modules/manager/npm/post-update/npm.ts

```diff
@@ -176,3 +205,66 @@ export async function generateLockFile(
   }
   return { lockFile };
 }
+
+export function divideWorkspaceAndRootDeps(
+  lockFileDir: string,
+  lockUpdates: Upgrade[]
+): {
+  lockRootUpdates: Upgrade[];
+  lockWorkspacesUpdates: Upgrade[];
+  workspaces: Set<string>;
+  rootDeps: Set<string>;
+} {
+  const lockRootUpdates: Upgrade[] = []; // stores all upgrades which are present in root package.json
+  const lockWorkspacesUpdates: Upgrade[] = []; // stores all upgrades which are present in workspaces package.json
+  const workspaces = new Set<string>(); // name of all workspaces
+  const rootDeps = new Set<string>(); // packageName of all upgrades in root package.json (makes it check duplicate deps in root)
+
+  // divide the deps in two categories: workspace and root
+  for (const upgrade of lockUpdates) {
+    if (
+      upgrade.isLockfileUpdate &&
+      upgrade.managerData?.workspacesPackages?.length &&
+      is.string(upgrade.packageFile)
+    ) {
+      const workspacePatterns = upgrade.managerData.workspacesPackages; // glob pattern or directory name/path
+      const packageFileDir = trimSlashes(
+        upgrade.packageFile.replace('package.json', '')
+      );
+
+      // workspaceDir = packageFileDir - lockFileDir
+      const workspaceDir = trimSlashes(packageFileDir.replace(lockFileDir, ''));
+
+      if (is.emptyString(workspaceDir)) {
+        lockRootUpdates.push(upgrade);
+        rootDeps.add(`${upgrade.packageName!}@${upgrade.newVersion!}`);
```

Comment:
create an utility function to generate this: `${upgrade.packageName!}@${upgrade.newVersion!}`.

It is often used in your changes.